{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "Algorithms_for_DNA Sequencing.ipynb",
      "provenance": [],
      "toc_visible": true,
      "authorship_tag": "ABX9TyNQUgs8KCzOWqzE+d8JghdP",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/xie186/Algorithms_for_DNA-Sequencing/blob/main/Algorithms_for_DNA_Sequencing.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "3QWthScNM_Qu"
      },
      "source": [
        ""
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "t-Vrd-XWNDQ6"
      },
      "source": [
        "# Week1 "
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "g6FsMCNYNILz"
      },
      "source": [
        "## Welcome to Algorithms for DNA Sequencing\n",
        "\n",
        "Welcome to Algorithms for DNA Sequencing!\n",
        "\n",
        "This module we begin our exploration of algorithms for analyzing DNA sequencing data. We'll discuss DNA sequencing technology, its past and present, and how it works. We'll learn more about DNA and genomes and we'll make the connection that a DNA molecule can be represented as a string of characters. This observation is the key that unlocks the entire rest of the course! It allows us to take methods from computer science that were invented to work with generic strings (web pages, books, files on your hard drive) and apply them to DNA.\n",
        "\n",
        "We will also define the two computational problems that motivate our technical discussions for the rest of the course. The read alignment problem and theassembly problem. The rest of the course is divided into halves where we will discuss foundational algorithms and data structures for those two crucial problems.\n",
        "\n",
        "In the practical sessions and homeworks, you will be working with real genomes and real sequencing data, including some human data. And you will start to write software to analyze them. You will learn how to work with important file formats like FASTA and FASTQ, and you will look for some basic patterns in that sequencing data. Finally, we will discuss the naive exact matching algorithm a simple, flexible, but not very fast algorithm for finding patterns in text. Next module we'll discuss algorithms that are a bit more clever, and see some new data structures as well.\n",
        "\n",
        "Your first two assignments, Programming Homework 1 and the Module 1 Quiz are due this week. Please also be sure to read the syllabus for all of the details about course logistics.\n",
        "\n",
        "Ben & Jacob"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "nueHoDOzNEVI"
      },
      "source": [
        "Course Content\n",
        "Module 1: DNA sequencing, strings and matching: Why study this? DNA sequencers and how they work. How DNA can be represented as a string. Using Python to parse and manipulate real genome sequences and real DNA sequencing data.  Naive exact matching.\n",
        "\n",
        "Module 2: Preprocessing, indexing and approximate matching: Improving on naive exact matching with Boyer-Moore.  Preprocessing and indexing.  Indexing: grouping and ordering.  k-mers and k-mer indexes.  Approximate matching and the pigeonhole principle.\n",
        "\n",
        "Module 3: Edit distance, assembly, overlaps: Hamming and edit distance. Algorithms for computing edit distance. Dynamic programming. Global and local alignment. De novo assembly. Overlaps and overlap graphs.\n",
        "\n",
        "Module 4: Algorithms for assembly: Shortest common superstring and the greedy version. How repetitive DNA makes assembly difficult. De Bruijn graphs and Eulerian walks. How real assemblers work. The future of assembly. Wrap up."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "D8ITcYKLNDWE"
      },
      "source": [
        "Using data files with Python programs\n",
        "\n",
        "In practical sessions and programming assignments, you will sometimes work with data files we provide.  These might be reference genomes in FASTA format (e.g. lambda_virus.fa), sequencing reads in FASTQ format (e.g. SRR835775_1.first1000.fastq), or Python modules (e.g. kmer_index.py).  To use a file with your Python code, you have to (a) download the file to the same computer as your Python program; use a web browser or a command-line tool like wget or curl to download it, then (b) put the file in the same directory with your Python program.  Then you can open it from your Python program using the Python open function.\n",
        "\n",
        "If you use Jupyter, another way to do this (which we use in the practical sessions) is with a command that starts with !wget.  This is a shell command.  But if you are using try.jupyter.org note that the !wget command will not work (\"unable to resolve host address\").  In this case, we recommend you either (a) use www.wakari.io instead, or (b) use the \"upload\" button in the try.jupyter.org interface to upload the file to the server, then skip the command that begins with !wget.\n",
        "\n",
        "Best,\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "J_4pv1vkNDbF"
      },
      "source": [
        ""
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "y8kMvRkPPa17"
      },
      "source": [
        ""
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "OBVa85iXP0GQ"
      },
      "source": [
        "## An implementation of the naive exact matching algorithm:"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "_WJpyGz9PcmP"
      },
      "source": [
        "# naive exact matching algorithm\n",
        "# Returns a list of occurrences (offset)\n",
        "def naive(p, t):\n",
        "    occurrences = []\n",
        "    for i in range(len(t) - len(p) + 1):  # loop over alignments\n",
        "        match = True\n",
        "        for j in range(len(p)):  # loop over characters\n",
        "            if t[i+j] != p[j]:  # compare characters\n",
        "                match = False\n",
        "                break\n",
        "        if match:\n",
        "            occurrences.append(i)  # all chars matched; record\n",
        "    return occurrences"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "zukxrwtxP5Qw",
        "outputId": "e3c41cff-57cf-4f4d-d642-59e6a542531f"
      },
      "source": [
        "p = \"ACGT\"\n",
        "ref = \"ACGTTCAGACGT\"\n",
        "naive(p, ref)"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "[0, 8]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 3
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "g2egjYcjQHTH"
      },
      "source": [
        "## Takes a DNA string and returns its reverse complement:\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "upi0EfSjPvOr"
      },
      "source": [
        "def reverseComplement(s):\n",
        "    complement = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A', 'N': 'N'}\n",
        "    t = ''\n",
        "    for base in s:\n",
        "        t = complement[base] + t\n",
        "    return t\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 35
        },
        "id": "VDDKUwlhQO8q",
        "outputId": "b53bd789-2e90-4b21-e0c4-431f90d0679b"
      },
      "source": [
        "p = \"ACGT\"\n",
        "reverseComplement(p)"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            },
            "text/plain": [
              "'ACGT'"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 6
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "MLP3Of9iQaBg"
      },
      "source": [
        ""
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "mPNkRgo2QY6P"
      },
      "source": [
        ""
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "BUadno5IQfy4"
      },
      "source": [
        "def readGenome(filename):\n",
        "    genome = ''\n",
        "    with open(filename, 'r') as f:\n",
        "        for line in f:\n",
        "            # ignore header line with genome information\n",
        "            if not line[0] == '>':\n",
        "                genome += line.rstrip()\n",
        "    return genome"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "lgchrttBQ8Kp"
      },
      "source": [
        "## Programming homework 1"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "D1f-NRGIRNBq",
        "outputId": "ab7c3ab5-9de9-49ec-f814-04c459fce8ec"
      },
      "source": [
        "# Phi-X genome\n",
        "!wget --no-check https://d28rh4a8wq0iu5.cloudfront.net/ads1/data/lambda_virus.fa\n"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "--2020-11-26 22:01:49--  https://d28rh4a8wq0iu5.cloudfront.net/ads1/data/lambda_virus.fa\n",
            "Resolving d28rh4a8wq0iu5.cloudfront.net (d28rh4a8wq0iu5.cloudfront.net)... 54.192.19.127, 54.192.19.137, 54.192.19.199, ...\n",
            "Connecting to d28rh4a8wq0iu5.cloudfront.net (d28rh4a8wq0iu5.cloudfront.net)|54.192.19.127|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 49270 (48K) [application/octet-stream]\n",
            "Saving to: ‘lambda_virus.fa’\n",
            "\n",
            "lambda_virus.fa     100%[===================>]  48.12K   205KB/s    in 0.2s    \n",
            "\n",
            "2020-11-26 22:01:51 (205 KB/s) - ‘lambda_virus.fa’ saved [49270/49270]\n",
            "\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "pPkMN7fiRTFP"
      },
      "source": [
        "phix_genome = readGenome('lambda_virus.fa')"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "U6f7JDN_R7Sg"
      },
      "source": [
        "# naive exact matching algorithm that is strand-aware\n",
        "# Instead of looking only for occurrences of P in T, additionally look for occurrences of the reverse complement of P in T. \n",
        "def naive_with_rc(p, t):\n",
        "\tr = reverseComplement(p)\n",
        "\tif r == p:\n",
        "\t\treturn naive(p,t)\n",
        "\telse:\n",
        "\t\treturn naive(p,t) + naive(r,t)"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "XVCWo_YgjMes"
      },
      "source": [
        "### Question 1:\n",
        "\n",
        "How many times does \\verb|AGGT|AGGT or its reverse complement (\\verb|ACCT|ACCT) occur in the lambda virus genome? E.g. if \\verb|AGGT|AGGT occurs 10 times and \\verb|ACCT|ACCT occurs 12 times, you should report 22."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "c0_qhvCjkEP3",
        "outputId": "552127cc-8453-41fa-910c-5d6c1a77d4ef"
      },
      "source": [
        "len(phix_genome)"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "48502"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 98
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "MgQxybpASHO0",
        "outputId": "64cf4084-881a-4815-8fb5-06aec9c01020"
      },
      "source": [
        "len(naive_with_rc(\"AGGT\", phix_genome))"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "306"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 99
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "WDl7Nj1xjXS6"
      },
      "source": [
        "### Question 2\n",
        "\n",
        "How many times does \\verb|TTAA|TTAA or its reverse complement occur in the lambda virus genome?\n",
        "\n",
        "Hint: \\verb|TTAA|TTAA and its reverse complement are equal, so remember not to double count."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "8PvV6TgwSZRw",
        "outputId": "d628887c-a236-4dcb-bc06-11f909624f44"
      },
      "source": [
        "\n",
        "len(naive_with_rc(\"TTAA\", phix_genome))"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "195"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 100
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "-ZXFASpQTbtR"
      },
      "source": [
        "### Question 3\n",
        " \n",
        " What is the offset of the leftmost occurrence of \\verb|ACTAAGT|ACTAAGT or its reverse complement in the Lambda virus genome? E.g. if the leftmost occurrence of \\verb|ACTAAGT|ACTAAGT is at offset 40 (0-based) and the leftmost occurrence of its reverse complement \\verb|ACTTAGT|ACTTAGT is at offset 29, then report 29.\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "IkvHY4gQSp3Y",
        "outputId": "00ade650-33aa-41ba-ee2c-7fea97c8d841"
      },
      "source": [
        "naive_with_rc(\"ACTAAGT\", phix_genome)"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "[27733, 45382, 26028]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 101
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "hlD8ciCATXzJ"
      },
      "source": [
        "### Question 4 \n",
        "What is the offset of the leftmost occurrence of \\verb|AGTCGA|AGTCGA or its reverse complement in the Lambda virus genome?\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "tDnG9D0aS03j",
        "outputId": "11d93e8d-269e-45dd-83fb-437b74331d14"
      },
      "source": [
        "naive_with_rc(\"AGTCGA\", phix_genome)"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "[18005, 23320, 33657, 44806, 450, 1908, 2472, 41927, 45369]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 102
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "gQAe3aZHlNHD",
        "outputId": "fc69a13b-7945-46e4-989a-c06e7da2e90f"
      },
      "source": [
        "min(naive_with_rc(\"AGTCGA\", phix_genome))"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "450"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 103
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "blSbtXjHTSRi"
      },
      "source": [
        "### Question 5 \n",
        "As we will discuss, sometimes we would like to find approximate matches for P in T. That is, we want to find occurrences with one or more differences.\n",
        "\n",
        "For Questions 5 and 6, make a new version of the \\verb|naive|naive function called \\verb|naive_2mm|naive_2mm that allows up to 2 mismatches per occurrence. Unlike for the previous questions, do not consider the reverse complement here. We're looking for approximate matches for P itself, not its reverse complement.\n",
        "\n",
        "￼\n",
        "\n",
        "For example, \\verb|ACTTTA|ACTTTA occurs twice in \\verb|ACTTACTTGATAAAGT|ACTTACTTGATAAAGT, once at offset 0 with 2 mismatches, and once at offset 4 with 1 mismatch. So \\verb|naive_2mm('ACTTTA', 'ACTTACTTGATAAAGT')|naive_2mm(’ACTTTA’, ’ACTTACTTGATAAAGT’) should return the list \\verb|[0, 4]|[0, 4]."
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "5JyWVoCYTn8H"
      },
      "source": [
        "\n",
        "def naive_2mm(p, t):\n",
        "    \"\"\"\n",
        "    Naive exact matching algorithm\n",
        "    Neturns a list of occurrences (offset)\n",
        "    \"\"\"\n",
        "    occurrences = []\n",
        "    for i in range(len(t) - len(p) + 1):  # loop over alignments\n",
        "        match = True\n",
        "        mismatch_count = 0\n",
        "        for j in range(len(p)):  # loop over characters\n",
        "            if t[i+j] != p[j]:  # compare characters\n",
        "                mismatch_count+=1\n",
        "\n",
        "                if mismatch_count > 2:\n",
        "                \tmatch = False\n",
        "                \tbreak\n",
        "        if match:\n",
        "            occurrences.append(i)  # all chars matched; record\n",
        "    return occurrences"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "AKGo7hqqTx7a",
        "outputId": "6fd90cfa-6a5d-4cd1-a219-6fc99ef4eb32"
      },
      "source": [
        "len(naive_2mm(\"TTCAAGCC\", phix_genome))"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "191"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 104
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "kfyQG2iJT82a"
      },
      "source": [
        "### Question 6\n",
        "\n",
        "What is the offset of the leftmost occurrence of \\verb|AGGAGGTT|AGGAGGTT in the Lambda virus genome when allowing up to 2 mismatches?"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "E8zcdQVkT7nx",
        "outputId": "1331ed19-c2d7-49cc-ae31-13e6f8f3091d"
      },
      "source": [
        "\n",
        "naive_2mm(\"AGGAGGTT\", phix_genome)"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "49"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 105
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "IKO-qHd1UKd3"
      },
      "source": [
        "### Question 7\n",
        "\n",
        "Finally, download and parse the provided FASTQ file containing real DNA sequencing reads derived from a human:\n",
        "\n",
        "https://d28rh4a8wq0iu5.cloudfront.net/ads1/data/ERR037900_1.first1000.fastq\n",
        "\n",
        "Note that the file has many reads in it and you should examine all of them together when answering this question. The reads are taken from this study:\n",
        "\n",
        "Ajay, S. S., Parker, S. C., Abaan, H. O., Fajardo, K. V. F., & Margulies, E. H. (2011). Accurate\n",
        "\n",
        "and comprehensive sequencing of personal genomes. Genome research, 21(9), 1498-1505.\n",
        "\n",
        "This dataset has something wrong with it; one of the sequencing cycles is poor quality.\n",
        "\n",
        "Report which sequencing cycle has the problem. Remember that a sequencing cycle corresponds to a particular offset in all the reads. For example, if the leftmost read position seems to have a problem consistently across reads, report 0. If the fourth position from the left has the problem, report 3. Do whatever analysis you think is needed to identify the bad cycle. It might help to review the \"Analyzing reads by position\" video.\n",
        "\n",
        "1 point\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "OgKrKpD8TTOf"
      },
      "source": [
        ""
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "siPlEye1UTnz",
        "outputId": "aa1e5fed-21c9-4420-b10c-8eb2400f6a94"
      },
      "source": [
        "!wget https://d28rh4a8wq0iu5.cloudfront.net/ads1/data/ERR037900_1.first1000.fastq"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "--2020-11-26 20:49:18--  https://d28rh4a8wq0iu5.cloudfront.net/ads1/data/ERR037900_1.first1000.fastq\n",
            "Resolving d28rh4a8wq0iu5.cloudfront.net (d28rh4a8wq0iu5.cloudfront.net)... 54.192.19.127, 54.192.19.199, 54.192.19.103, ...\n",
            "Connecting to d28rh4a8wq0iu5.cloudfront.net (d28rh4a8wq0iu5.cloudfront.net)|54.192.19.127|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 241626 (236K) [application/octet-stream]\n",
            "Saving to: ‘ERR037900_1.first1000.fastq’\n",
            "\n",
            "ERR037900_1.first10 100%[===================>] 235.96K   395KB/s    in 0.6s    \n",
            "\n",
            "2020-11-26 20:49:20 (395 KB/s) - ‘ERR037900_1.first1000.fastq’ saved [241626/241626]\n",
            "\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "TeOkV_LLUuWg"
      },
      "source": [
        "\n",
        "def readFastq(filename):\n",
        "    \"\"\"\n",
        "    Read fastq files\n",
        "    \"\"\"\n",
        "    seqs = []\n",
        "    quals = []\n",
        "    with open(filename) as fh:\n",
        "        while True:\n",
        "            fh.readline()  # skip name line\n",
        "            seq = fh.readline().rstrip()  # read base sequence\n",
        "            fh.readline()  # skip placeholder line\n",
        "            qual = fh.readline().rstrip() # base quality line\n",
        "            if len(seq) == 0:\n",
        "                break\n",
        "            seqs.append(seq)\n",
        "            quals.append(qual)\n",
        "    return seqs, quals"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "59YI76vSUrQK"
      },
      "source": [
        "sequences, base_quals = readFastq('ERR037900_1.first1000.fastq')\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "15a8hzk4VwHr",
        "outputId": "de590619-7a55-4092-c00e-affca307f96e"
      },
      "source": [
        "len(sequences[1])"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "100"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 33
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "gcPR7DPyVTlh",
        "outputId": "3dc4ccb5-1b24-4a7b-cd55-86693c58d7b7"
      },
      "source": [
        "base_quals[0:2]\n",
        "\n"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "['HHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHGFHHHFHFFHHHHHGHHFHEH@4#55554455HGFBF<@C>7EEF@FBEDDD<=C<E',\n",
              " 'HHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHCHHHHEHHBA#C>@54455C/7=CGHEGEB;C############']"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 32
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "CDZn_C7WXExw",
        "outputId": "7e59c752-8479-42af-c7f4-d0920f9feb53"
      },
      "source": [
        ""
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "0\n",
            "1\n",
            "2\n",
            "3\n",
            "4\n",
            "5\n",
            "6\n",
            "7\n",
            "8\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "BvKw3MgPW4HZ"
      },
      "source": [
        "def averageQuals(quals):\n",
        "    counter = {}\n",
        "    qual_sum = {}\n",
        "    average_qual = {}\n",
        "    for qual in quals:\n",
        "        for i in range(len(qual)):\n",
        "            base_qual = qual[i]\n",
        "            phred_score = ord(base_qual) - 33\n",
        "            #print(phred_score)\n",
        "            if i not in qual_sum:\n",
        "                qual_sum[i] = phred_score\n",
        "            else:\n",
        "                qual_sum[i] += phred_score\n",
        "            if i not in counter:\n",
        "                counter[i] = 1\n",
        "            else:\n",
        "                counter[i] += 1\n",
        "    for ele in qual_sum:\n",
        "        #print(\"%d %d  %d\" % (ele, qual_sum[ele], counter[ele]))\n",
        "        average_qual[ele] = qual_sum[ele]/counter[ele]\n",
        "    return average_qual\n",
        "             \n",
        "\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "ngsh5P66YhOb"
      },
      "source": [
        "average_qual = averageQuals(base_quals)\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Z7byMxS2bPRD"
      },
      "source": [
        "import matplotlib.pyplot as plt\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "z9y-Suuneq2a",
        "outputId": "e7a6cda9-22f4-49be-c1b2-cddefa2bc238"
      },
      "source": [
        "\n",
        "average_qual.values()"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "dict_values([38.435, 38.46, 38.4, 38.428, 38.417, 38.382, 38.362, 38.364, 38.325, 38.343, 38.194, 38.154, 38.089, 38.152, 38.153, 38.099, 38.055, 38.041, 38.067, 38.15, 38.087, 37.989, 38.009, 37.909, 38.028, 37.576, 37.514, 37.503, 37.578, 37.552, 37.629, 37.555, 37.548, 37.433, 37.402, 37.343, 37.198, 37.126, 37.158, 36.904, 36.742, 36.677, 36.645, 36.493, 36.509, 36.375, 36.205, 35.974, 35.847, 35.673, 34.832, 34.739, 34.399, 34.402, 34.025, 33.025, 32.882, 32.741, 32.573, 32.224, 30.805, 30.666, 30.343, 30.131, 29.463, 17.891, 4.526, 17.881, 17.865, 17.852, 17.466, 17.131, 17.134, 17.098, 16.852, 23.745, 23.958, 23.221, 22.902, 22.366, 20.904, 20.584, 20.383, 19.767, 19.095, 18.061, 17.757, 17.152, 16.972, 16.286, 14.811, 14.679, 14.31, 14.152, 13.734, 12.538, 12.278, 12.179, 11.676, 11.561])"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 77
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "SY8_zot7YxxK",
        "outputId": "e758fab7-de62-4cc4-ccec-785c601385e0"
      },
      "source": [
        "\n",
        "average_qual.keys()"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "dict_keys([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99])"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 78
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 285
        },
        "id": "ApeoKujMek85",
        "outputId": "c1b87e3f-4ca5-438d-89bc-3ebeb9716ac9"
      },
      "source": [
        "plt.plot(list(average_qual.keys()),  list(average_qual.values()))\n"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "[<matplotlib.lines.Line2D at 0x7fac9b5aad68>]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 81
        },
        {
          "output_type": "display_data",
          "data": {
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXAAAAD7CAYAAABzGc+QAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+WH4yJAAAgAElEQVR4nO3de3xcdZ3/8ddnbslMmmuTtmlK74VaSi9QqpWCXBdBUFB3H+Iui8huZdcLqOsu6v7U1dVdVxEvv11ckKs3VC6WHyKKXOSitDS0lNJCW0ov6TVtkibNbSYz398fMwlpm7S5THJ6Zt7PxyOPZs5MMp/T077zzed8z/macw4REfGfgNcFiIjI0CjARUR8SgEuIuJTCnAREZ9SgIuI+JQCXETEpwYc4GYWNLPVZvZI5vE0M1thZpvN7BdmFhm5MkVE5EiDGYHfAGzo9fibwC3OuZlAI3BdNgsTEZFjs4FcyGNmk4B7gK8DnwEuB+qBCc65LjNbAnzFOXfxsb5PZWWlmzp16rCLFhHJJ7W1tfudc1VHbg8N8Ou/C/wzUJx5PBZocs51ZR7XATV9faGZLQOWAUyePJlVq1YNpm4RkbxnZtv62n7cFoqZXQbsc87VDuWNnXO3OecWOecWVVUd9QNERESGaCAj8LOA95rZpUAhUAJ8Dygzs1BmFD4J2DlyZYqIyJGOOwJ3zn3eOTfJOTcV+BDwpHPur4GngA9mXnYNsHzEqhQRkaMMZx74vwCfMbPNpHvid2SnJBERGYiBnsQEwDn3NPB05vMtwOLslyQiIgOhKzFFRHxKAS4i4lODaqF4Zc2OJjbubaGts4vWeBKA8SWFTCwtpLwoQls8yaHOLtrjSUIBIxwKEAkGKC4MURoNUxINE4sECQf180pEcocvAvyB2jp+/EKf89gHJRQwCsNBDOi+/nRMQYiyWJjSaBiAjq4UnYkk0yqLOPeUKs45uYrq0uiw31tEJNsGdCl9tixatMgN5UrM+pZOOhJJigpCFBUEcQ72HOxgV1M7Te0JYpEgxYUhCsNBkilHIumId6Vo7khwsD1Bc3uCjkSS9kSS9ngKh8MwHI5DHV00tSdoaotjGAXh9Oj91V3N7GnuAKCiKEJ5LEx5LMK4kgJqyqJMKo8RDBg7m9rTdbQlen4oRIIBxpUUMKGkkHHFBYwdU9DzPSKhAJFQgFAgQFu8i9bOJAfbE7y5/xBv1Leyo6GNsliEmrJCasqjzBpXzKzxYygIBbN4JETET8ys1jm36MjtvhiBVxUXHLVtamURUyuLRuw9nXNs3HuIP27cx/aGNhpbEzS0xnltTwtPbNhHZ1cKSI/qq8sKKY9FMDMM6EgkeWl7Iw2t8UG9ZyQUYFJ5lINtCQ70+tpQwJheVURBKEhHIklHV5LxxYWcOrGEt1WXcLA9Qe22RtbsaGJuTSm3/NUCSmPhbP51iMgJyBcj8BONc479h+IkU46q4gKCAevzdZ1dSfYfitNwKM6B1k6a2hLEu1J0JlMkkyliBSGKIiGKC0NMHVtETXm053u1x5PUNbbx+t4W1u9qZuPeFpIpR2E4SCQUYGdjOxt2N/ecE5g6NsapE0v5/fo9TK6IcddHFjN5bGzU/k5EZOT0NwJXgPtYKuXY3tDGmMIQlWPSv6W8sOUAH/txLaGA8aXL5zCjagw1ZVHKYmHM0j8ckinHrqZ2th5o5cChOLFIkKKC9C9j2w60sfVAK/UtnVSXFjJ1bBGTyqMUhAMEA+n20vSqIgrDaumIjBYFeB7ZUn+Ij979IlsPtPVsM4NwIEA4aMSTKRLJ/o97JBSgsijC3pZOkqmjXxcOGnOqS5h/UhnjSwopLgxRUhimqriA6tJCqkujRCMKeJFs8XUPXAZnetUYHrvxHDbubWFXUzs7mzpoaouTSDq6kinCoQBTx8aYMraIquIC2uNJWju7SDmYPDZGdUkhgYCRSKYyX99OIulIplK0xZO8uquZ1dsbeaC2rqeFc6STx4/h7FlVnD2rkvmTyg77DUBEskMjcBmWzq4kLR1dNLcn2NvcyZ7mduoa2lm5tYEVbzYQz5zsLS4MMbkiRkEoQHPm9dMqi7j+3Bmce3KVwl3kGDQClxFREApSMCZI5ZgCpleNOey5jkSSF7c28PqeFrY3tLHtQBvJlGNCaSFFkRDPbd7PtXe9yNyaEq5bOo3zZ4/vmY8vIsenEbh4Jt6V4qHVddz69BtsPdBGKGC8fXoFZ82s5KTyGJPKo0woLaSkMH0lrUbpkq90ElNOWKmUY/WOJh5fv5fH1+/hjfrWo14TDBjjigs4a2Yl550yjsXTKjCjp0VTXVqogJecpQAX32jpSLCzqZ2dje3sbe6kpSNBc0eCrQfaeG7Tfg62J476mnHFBSyZMZZ3zhjLebPHMa640IPKRUaGeuDiG8WFYWZPCDN7QslRz3UlU6zZ0cTLdQcJB42CUIDOrhQvbm3k+c0HWL5mF2awaEo5F586gfefPomKoogHeyEy8jQCl5zhnOP1vS38bt1efrtuN6/taaEgFODKhTVce9Y0TplQ7HWJIkOiForknY17W7jr+a08+FIdnV0pplcWcebUChZPq+Cck6v6vMeOyIlIAS55q7E1zgMv1fHClgOsfLOB5o4uAgaLp1Vw6WnVfOD0ST23EhA5ESnARUjPeFm/u5nfr9/Lo6/sZvO+Q5x7ShV3feRMzWKRE1Z/AX7cJWrMrNDMVprZy2b2qpn9W2b73Wb2ppmtyXwsGInCRbIpEDDm1pTymYtO5g+feRdfvPRtPP16PY+s3e11aSKDNpA1xjqB851z84EFwLvN7B2Z5z7nnFuQ+VgzYlWKjJCPLp3GaTWl/Nv/W9/n9ESRE9lxA9ylHco8DGc+Rq/vIjKCggHjP95/Gg2tnXzzsde8LkdkUAa0yq+ZBc1sDbAPeNw5tyLz1NfNbK2Z3WJmfZ7SN7NlZrbKzFbV19dnqWyR7JlbU8q1Z03jZyu28+BLdew52MFonhsSGapBncQ0szLgIeCTwAFgDxABbgPecM599Vhfr5OYcqJq7ezi0u8/y7bMPdTLYmEunjOBmy6ZTbkuBBKPZeVKTOdck5k9BbzbOfftzOZOM7sL+Kcs1CniiaKCEI9+6mxe2XmQ13Y388rOZh54qY4nXtvLly8/lcvmVWuWipxwjjsCN7MqIJEJ7yjwe+CbQK1zbrel/1XfAnQ452461vfSCFz8ZP2uZm56cC1r6w5SU5Zer9ThmDWumK9dMZeasqjXJUqeGPI8cDObB9wDBEn3zH/pnPuqmT0JVAEGrAGu73Wys08KcPGbrmSKn7ywjdU7mugefz++fm/m5Oc83jOv2tP6JD/oQh6RLNl2oJUb7lvDmh1N/O2SKXz1fXO9Lkly3JAv5BGRw00ZW8Svrl/CNUumcO+ft/GnN/Z7XZLkKQW4yBCEgwE+f+nbqC4t5L8ee13TDsUTCnCRISoMB7nxwlms2dHE79fv9bocyUMKcJFh+MDpk5heVcS3f/c6yZRG4TK6FOAiwxAKBvjsRaewad8hfr16p9flSJ7RTZBFhumSuROYW1PCNx7dwMa9LSydVcmZUysoDAe9Lk1ynKYRimTBq7sO8rVH1lO7rZFE0lEQCnDm1AqWzqpk6cxKTp1Yois5Zcg0D1xkFLR2drHyzQae3bSf5zfv5/W9LQDUlEW5aM54Lpk7gcXTKhTmMigKcBEP7Gvu4I8b6/ndq3t5ZlM98a4UF75tPN+4ci7jSgq9Lk98QgEu4rHWzi5+umIbN/9+IwWhAF++/FTef3qNRuNyXLoSU8RjRQUhlp0zg9/ecDanTCjms796mcc1f1yGQQEuMsqmV43hvmVLmFBSyC9e3OF1OeJjCnARDwQDxhULa3h6Yz31LZ1elyM+pQAX8cgHz6ghmXIsX6MLgGRoFOAiHpk5rpj5J5Vxf22dboYlQ6IAF/HQB0+v4bU9Lby6q9nrUsSHFOAiHrp8/kQiwQAPvFTndSniQwpwEQ+VxSJcOGccD6/ZRSKZ8roc8RkFuIjHPnD6JA60xrnmzpXc/swW1u9qVk9cBkR3IxTx2LmnjOMT583kt+t28/VHNwDw9mkVfP3KucwcV+xxdXIiG8iq9IXAM0AB6cC/3zn3ZTObBtwHjAVqgaudc/FjfS9dSi9ybHsOdvDbdbu55fGNtCeS/P3Z0/nH82YypkBjrXw25HuhWPpGDUXOuUNmFgaeA24APgM86Jy7z8x+CLzsnLv1WN9LAS4yMPsPdfIfj77GAy/VURQJ8r6FNXx48WTm1pR6XZp4YMj3QnFphzIPw5kPB5wP3J/Zfg9wRZZqFcl7lWMKuPmv5rP842fx7rnVPFBbx2U/eI4vLV/ndWlyAhnQSUwzC5rZGmAf8DjwBtDknOvKvKQOqBmZEkXy1/yTyrj5r+az8gsXcvU7pnDvn7fx85XbvS5LThADCnDnXNI5twCYBCwGZg/0DcxsmZmtMrNV9fX1QyxTJL+VxsJ85b2ncvasSr68/FXW7GjyuiQ5AQxqGqFzrgl4ClgClJlZ95mVSUCfN3Rwzt3mnFvknFtUVVU1rGJF8lkwYHz/QwsZV1LAP/yklv2HdBOsfHfcADezKjMry3weBS4CNpAO8g9mXnYNsHykihSRtPKiCP979Rk0tMa5+feve12OeGwgc5OqgXvMLEg68H/pnHvEzNYD95nZvwOrgTtGsE4RyTh1Yimzq0vY1dThdSniseMGuHNuLbCwj+1bSPfDRWSUFUWCtMW7jv9CyWm6lF7Eh2KRIG3xpNdliMcU4CI+FIuEFOCiABfxo6KCIK2daqHkOwW4iA9FwyHaNQLPewpwER8qKgjSGu/SbWfznAJcxIdikRApB51dWgQinynARXwoFgkC6ERmnlOAi/hQd4DrRGZ+U4CL+FBRZoEHjcDzmwJcxIei3SNwXY2Z1xTgIj5UFEmPwDWVML8pwEV8SD1wAQW4iC9pFoqAAlzEl3QSU0ABLuJLb43A1ULJZwpwER+KZU5itnZqBJ7PFOAiPhQMGAWhAG0JjcDzmQJcxKeKCkK0aQSe1xTgIj4VDQd1IU+eU4CL+FRRQVAX8uQ5BbiIT8UiIVoV4HntuAFuZieZ2VNmtt7MXjWzGzLbv2JmO81sTebj0pEvV0S6xSJB2nQlZl4LDeA1XcBnnXMvmVkxUGtmj2eeu8U59+2RK09E+hOLhGhsa/e6DPHQcQPcObcb2J35vMXMNgA1I12YiBxbugeuEXg+G1QP3MymAguBFZlNnzCztWZ2p5mV9/M1y8xslZmtqq+vH1axIvKWWCSoHnieG3CAm9kY4AHgRudcM3ArMANYQHqEfnNfX+ecu805t8g5t6iqqioLJYsIpFso6oHntwEFuJmFSYf3T51zDwI45/Y655LOuRRwO7B45MoUkSMVRYK0JZJamT6PDWQWigF3ABucc9/ptb2618uuBNZlvzwR6U+sIIRz0JHQyvT5aiCzUM4CrgZeMbM1mW1fAK4yswWAA7YCHxuRCkWkT7Fey6p1L7Em+WUgs1CeA6yPpx7NfjkiMlCxE2RZtc37DvGzFdt5cHUd7184iS9dPsfTevLJQEbgInICKvJ4YWPnHH9/by1/2LCXcNCYUTWGO59/kyUzxnLRnPGe1JRvdCm9iE/1rEzv0R0J2+JJ/rBhL5fPn8ifP38Byz9xFnOqS/iXB9ayr7nDk5ryjQJcxKe6l1XzqoXS0BoH4OxZlVSOKaAgFOT7Vy2kLd7FZ3/1MqmUZseMNAW4iE/FPG6hNLalA7wiFunZNnPcGL502ak8u2k///6bDXQkdKHRSFKAi/hU90lMr9bFbGxLAFBeFD5s+1WLT+LDb5/Mnc+/yV/c8gxPvbbPi/LyggJcxKeKehY29maU25hpoZT3GoEDmBnfuPI0fvp3byccNK69+0X+6Vcvk1RLJesU4CI+Fcv0wL1aVq27hXJkgHc7a2Ylv73hHD5x3kzur63jcwrxrNM0QhGfioY97oG3xgkYlETD/b4mEgrwTxefQkEowM2PbyQYML75gXkEAn1dWiKDpQAX8algwCgMBzxroTS0xSmLRQgOIIw/ecEsks7x3T9sojQa5l8v08U+2aAWioiPFUVCnp7ELIv1P/o+0g0XzOIDp0/i3he20dyRGMHK8ocCXMTHopGgdz3w1vhhUwiPx8y4eskU4l0pHntlzwhWlj8U4CI+VhQJeTgPPEHZIAIcYP6kUqZVFvHg6roRqiq/KMBFfCxWEPR0GmFF0cBbKJAehV+5sIYXtjSws0nreQ6XAlzEx2IRbwLcOUdDW5zyosGNwAGuWJBeUnf5mp3ZLivvKMBFfCwWCdHqwbJq7Ykk8a5Uv3PAj2Xy2BhnTCnnoZd2ajWhYVKAi/hYUSRIuwf3G+m+kdVgTmL2duXCGjbtO8T63c3ZLCvvKMBFfCwaCXlyO9mmzH1QBjONsLf3nFZNOGg89JLaKMOhC3lEfKwoEvRkHnjPCHwIPXCA8qIIF75tPHf/aSuhYIBPXTCz5+ZcMnAagYv4WKwgRHsiOer33u6+D8pgpxH29o0rT+PKhTX88I9vcOHNf+Thl3eRSGqB5sFQgIv4WCwSTK9M3zW6bZTGYY7AIT0K/9Zfzuf+65dQEg3zqZ+v5h3feIKvPbKezfsOZavUnHbcADezk8zsKTNbb2avmtkNme0VZva4mW3K/Fk+8uWKSG9FHi2r1tCWwAxKj3Ejq4FaNLWCRz65lDs/sogzp1Zw75+3cun3nmXFlgPDLzTHDWQE3gV81jk3B3gH8HEzmwPcBDzhnJsFPJF5LCKjyKuV6Zva4pRGwwO6kdVAhIIBzp89nh9efQbP33Q+J1VE+dhPanlzf2tWvn+uOm6AO+d2O+deynzeAmwAaoD3AfdkXnYPcMVIFSkiffNqWbWGQd4HZTDGFRdy10cWEzDj2rtW9rRr5GiD6oGb2VRgIbACGO+c2515ag8wvp+vWWZmq8xsVX19/TBKFZEj9SzqMMoB3tgWH/IUwoGYPDbG7X97BrsOdvCxH9cS79LJzb4MOMDNbAzwAHCjc+6w2fcufTlVn6fBnXO3OecWOecWVVVVDatYETmcVz3wxtbEsE5gDsQZUyr49l/OZ+XWBr72yPoRfS+/GlCAm1mYdHj/1Dn3YGbzXjOrzjxfDWjlUpFRFvVoXczGzGIOI+298yey7Jzp/PiFbTxQqzsYHmkgs1AMuAPY4Jz7Tq+nHgauyXx+DbA8++WJyLEUebQyfWNbfMRH4N3++eJTWDJ9LF946BXW7Tw4Ku/pFwMZgZ8FXA2cb2ZrMh+XAv8JXGRmm4ALM49FZBTFCrpPYo7eCLw9nqQjMbQbWQ1FKBjgBx9eSEVRhOt/UkuLVvPpMZBZKM8558w5N885tyDz8ahz7oBz7gLn3Czn3IXOuYbRKFhE3vLWNMLRG4E39KxGP3InMY9UOaaAH1y1kLrGdu56fuuove+JTldiivhYz8r0o3gSs3ta31DuBT4ci6ZWcNGc8dz+7BYOtmkUDgpwEV8LBoxoeHRvaNXYMwIf3QAH+PSFJ9PS0cWPntsy6u99IlKAi/jcaK/K89adCEevhdJtzsQSLj1tAnc+96Yu8EEBLuJ7o70uZve9wL0YgQPceOHJtCWS/O8zGoUrwEV8rmiUl1XrHoFn40ZWQ3Hy+GLeO38i9/xpK7XbGj2p4UShO6iL+Fy0n2XVarc1ct/K7URCAQpCQSKhAJGgEQ4GqC6LcuXCmiHdjKr7RlahoHfjv89cdDIvbDnAB279E1csmMi/XDKb6tKoZ/V4RQEu4nOxSLDPEfh9K7fzwEt1VBRF6Eyk6OxKEe+1YMKvVu3gux9aMOjga2hLjOoUwr5MGVvEk589l/95ejO3P/smv3t1L/979Rmcc3J+3a5DAS7ic9FwiMbWo6fVtSeSTK1MB1035xxdKcfDa3bxf5av45LvPcu3Pjifi+b0eS+6PjW2xkd9CmFfigpCfO7i2XzozMks+3Et//CTWn55/RJOnVjqdWmjRj1wEZ+LRoJ09NFC6Ugke+aJdzNLt1A+cMYkHvnkUmrKovz9vav49C/WcOBQ54Der7Et7tkJzL6cVBHj7mvPpCQa5qN3v8iupnavSxo1CnARn4uGA33OQmnvI8B7m141hgf/8Z186oJZPLJ2Fxd+54/85IVtPLupnmc2pj9Wb29kS/0hmtremrLX2HpiBTjA+JJC7rr2TNo6k3z07hfZdiA/FoJQC0XE56Lhvk9itseTFBUc+794QSjIZy46mcvnVXPTg6/wr79e1+9rz55VyfXvmkFjW8KTOeDHM3tCCT+8+gyuvetF3vWtp5k3qZTL5lXz4bdPYcxx/h78Kjf3SiSPFPYzC6U9kWLsmP5H4L3NGl/Mrz62hLU7D9KVTGEGzkFLRxdN7XG27m/jZyu389c/WgEMbzX6kXTWzEqe/ty5PLJ2F4+s3c03Hn2N5zYf4K6PnJm15d9OJApwEZ+LhoPEu1IkU+6wkOqrB34sgYCx4KSyfp//h3Nn8NDqnTy0eifvnDF2WDWPpIllUZadM4Nl58zgpyu28cWH1vGDJzdx44Une11a1inARXyuO6Q7Eoe3TNriXYMK8OMpDAe5avFkrlo8OWvfc6R9ePFkarc28r0nNrFwcjnvyrFphjqJKeJz3avyHNlGaY8ne57LV2bG1688jZPHFXPjfavZmWMzVBTgIj7XPcpuP2ImSkciRWEWR+B+FY0EufVvTqcjkeK7j2/0upysUoCL+Fz3KLv3XPCuZPqqy1iej8C7Ta8aw3vmVfPYuj19zpn3KwW4iM/1jMB7BVNHV+qw5wSuWFBDS2cXT2zInfXXFeAiPtcd0r0v5ulupxRqBN5jyYyxjCsu4NdrdnpdStYowEV8rrCPk5jdbQKNwN8SDBjvnT+Rp1/flzOLQRw3wM3sTjPbZ2brem37ipntPGKVehHxQM80wt4jcAV4n65YWEMi6fjNK7u9LiUrBjICvxt4dx/bb+m9Sn12yxKRgeqrB97dQolG9Et2b6dOLGHWuDEsz5E2ynGPrnPuGaBhFGoRkSHoax54dz9c0wgPZ2ZcsbCGF7c2sqOhzetyhm04P54/YWZrMy2W8v5eZGbLzGyVma2qr68fxtuJSF8K+5gHrh54/947fyIAX3joFR5aXcfug/69uGeoAX4rMANYAOwGbu7vhc6525xzi5xzi6qqcusyVpETQayPeeA9PXDNQjnKSRUxPn7eDNbWHeTTv3iZJf/xJH/zoxW+vI/4kALcObfXOZd0zqWA24HF2S1LRAYqHAwQClifPfBYWLc76svnLp7N6v9zEb/51FI+d/EpvLS9kYu/+4zveuNDOrpmVu2c6z6NeyXQ/02ERWTERcNB2uNvrXfZHeaFOonZr0DAOHViKadOTN83/NO/WMMN963hvpU7uOS0CZw/exyTymNel3lMA5lG+HPgz8ApZlZnZtcB/2Vmr5jZWuA84NMjXKeIHMOR9wRXD3xwpowt4pcfW8JNl8xmT3MHX1r+Kku/+RQfvv0F9jV3eF1ev447AnfOXdXH5jtGoBYRGaL0CPytlenbNQtl0ELBANe/awbXv2sGW+oP8dire/jBE5u57AfPcevfnM4ZUyq8LvEo+v1KJAccuaxaeyJJOJhewFgGb3rVGP7x3Jk89PF3Eo0E+dBtL/CLF7d7XdZRdHRFckC6hXJ4D1yj7+GbPaGEhz++lMXTKvjiQ+vYe4K1UxTgIjkgGg4cfil9fHDLqUn/SmNhvnHlaSSd494/b/W6nMMowEVyQF8tFM0Bz54pY4u46G3j+emK7UctnOElBbhIDogeMQtFI/Dsu27pNJraEjy4us7rUnoowEVyQDQcOmxkqBF49i2eVsHcmhLufO5NUinndTmAAlwkJ0QjgcMupe9IaASebWbGdUun8UZ9K89sOjHu66QAF8kBffbAFeBZ957TJjKuuID/efoNWju7jv8FI0wBLpIDugPcufSv9u3xpJZTGwGRUICPnzeTlW82sPSbT/LfT22mpSPhWT26041IDiiMBHEOOrtSFIaDdCRSGoGPkGveOZV5k0r5/hOb+NbvXucHT25i9oQS5kws4YzJ5Vy5sIZAwEalFgW4SA6I9roneGFmNK4AHzkLJ5dz17WLWVvXxPI1u1i/q5nfrN3Nz1Zs5/nN+/mvD84jNApXwSrARXJA72XVyoG2eJdmoYyCeZPKmDepDADnHP/3yc3c/PhG2hNJvvehhURCIxviCnCRHNB7WbVUytGRSOlS+lFmZnzyglnECkJ87ZH1HLrnRf7u7OksmFRGaSw8Iu+pABfJAb2XVevsSt8TRS0Ub1y3dBqxSJAvLV/Hs5v2AzCjqoh/v+I0lswYm9X3UoCL5IDey6p1TyeMqYXimasWT+ayedWsrTvImh1NrN7eSOWYSNbfRwEukgN698DbtZjDCaG4MMxZMys5a2bliL2H5oGL5IDeLZSexRw0As95CnCRHND7JKaWU8sfCnCRHNB7HrhaKPlDAS6SA3r3wNsyLZSoVqTPeQNZlf5OM9tnZut6basws8fNbFPmz/KRLVNEjqV3C0ULGuePgfyIvht49xHbbgKecM7NAp7IPBYRjxRkrvjriKsHnk+OG+DOuWeAhiM2vw+4J/P5PcAVWa5LRAbBzHruSNjTA9cslJw31CbZeOfc7szne4Dx/b3QzJaZ2SozW1Vff2LcBF0kF8Uyy6p1t1BiYV3mkeuGfZbDpW9A3O/6Qs6525xzi5xzi6qqqob7diLSj8JwkPZ4qmcEXqiTmDlvqEd4r5lVA2T+3Je9kkRkKKKRIB2ZeeABg8go3M5UvDXUI/wwcE3m82uA5dkpR0SGqqcHnlmR3mx0FhUQ7wxkGuHPgT8Dp5hZnZldB/wncJGZbQIuzDwWEQ9Fw8GeC3l0AjM/HPcsh3Puqn6euiDLtYjIMBRGghxsT/SsyiO5T00ykRwRDQfo6B6BK8DzggJcJEf0ngeuFkp+UICL5Ihor3ngaqHkBwW4SI6IhkM9l9JrNZ78oAAXyRHRSOCtFopG4HlBAS6SI6LhIF0pR0tHlwI8TyjARXJEd9/7QGtcy6nlCQW4SI7onnkS70ppBLDKEAgAAAWoSURBVJ4nFOAiOaJ3aCvA84MCXCRHHBbgaqHkBQW4SI7o3ffWPPD8oAAXyRFqoeQfBbhIjuh98Y4u5MkPCnCRHNF71K0WSn5QgIvkiEKdxMw7CnCRHNE7tNUDzw8KcJEcoZOY+UcBLpIjDm+h6L92PtBRFskRwYARCaX/S+skZn5QgIvkkO7WiVoo+eG4ixofi5ltBVqAJNDlnFuUjaJEZGii4fTCxpqFkh+GFeAZ5znn9mfh+4jIMHVfwFMYUoDnA7VQRHJIYThIYThAIGBelyKjYLgB7oDfm1mtmS3r6wVmtszMVpnZqvr6+mG+nYgcSzQSVP87jww3wJc6504HLgE+bmbnHPkC59xtzrlFzrlFVVVVw3w7ETmWaFgBnk+GFeDOuZ2ZP/cBDwGLs1GUiAxNuoWiAM8XQz6JaWZFQMA515L5/C+Ar2atMhEZtI+8cyr7D3V6XYaMkuHMQhkPPGRm3d/nZ865x7JSlYgMydJZlV6XIKNoyAHunNsCzM9iLSIiMgiaRigi4lMKcBERn1KAi4j4lAJcRMSnFOAiIj6lABcR8SkFuIiIT5lzbvTezKwe2DbEL68E8vG2tfm43/m4z5Cf+52P+wyD3+8pzrmjbiY1qgE+HGa2Kh8XjMjH/c7HfYb83O983GfI3n6rhSIi4lMKcBERn/JTgN/mdQEeycf9zsd9hvzc73zcZ8jSfvumBy4iIofz0whcRER6UYCLiPiULwLczN5tZq+b2WYzu8nrekaCmZ1kZk+Z2Xoze9XMbshsrzCzx81sU+bPcq9rzTYzC5rZajN7JPN4mpmtyBzvX5hZxOsas83MyszsfjN7zcw2mNmSXD/WZvbpzL/tdWb2czMrzMVjbWZ3mtk+M1vXa1ufx9bSvp/Z/7Vmdvpg3uuED3AzCwL/TXrh5DnAVWY2x9uqRkQX8Fnn3BzgHaQXiZ4D3AQ84ZybBTyReZxrbgA29Hr8TeAW59xMoBG4zpOqRtb3gMecc7NJL4yygRw+1mZWA3wKWOScmwsEgQ+Rm8f6buDdR2zr79heAszKfCwDbh3MG53wAU56oeTNzrktzrk4cB/wPo9ryjrn3G7n3EuZz1tI/4euIb2v92Redg9whTcVjgwzmwS8B/hR5rEB5wP3Z16Si/tcCpwD3AHgnIs755rI8WNNegWwqJmFgBiwmxw81s65Z4CGIzb3d2zfB9zr0l4AysyseqDv5YcArwF29Hpcl9mWs8xsKrAQWAGMd87tzjy1h/RapLnku8A/A6nM47FAk3OuK/M4F4/3NKAeuCvTOvpRZmHwnD3WzrmdwLeB7aSD+yBQS+4f6279Hdth5ZsfAjyvmNkY4AHgRudcc+/nXHrOZ87M+zSzy4B9zrlar2sZZSHgdOBW59xCoJUj2iU5eKzLSY82pwETgSKObjPkhWweWz8E+E7gpF6PJ2W25RwzC5MO75865x7MbN7b/StV5s99XtU3As4C3mtmW0m3xs4n3Rsuy/yaDbl5vOuAOufciszj+0kHei4f6wuBN51z9c65BPAg6eOf68e6W3/Hdlj55ocAfxGYlTlbHSF94uNhj2vKukzv9w5gg3PuO72eehi4JvP5NcDy0a5tpDjnPu+cm+Scm0r6uD7pnPtr4Cngg5mX5dQ+Azjn9gA7zOyUzKYLgPXk8LEm3Tp5h5nFMv/Wu/c5p491L/0d24eBv83MRnkHcLBXq+X4nHMn/AdwKbAReAP4otf1jNA+LiX9a9VaYE3m41LSPeEngE3AH4AKr2sdof0/F3gk8/l0YCWwGfgVUOB1fSOwvwuAVZnj/WugPNePNfBvwGvAOuDHQEEuHmvg56T7/AnSv21d19+xBYz0LLs3gFdIz9IZ8HvpUnoREZ/yQwtFRET6oAAXEfEpBbiIiE8pwEVEfEoBLiLiUwpwERGfUoCLiPjU/wftS14Qq59XngAAAABJRU5ErkJggg==\n",
            "text/plain": [
              "<Figure size 432x288 with 1 Axes>"
            ]
          },
          "metadata": {
            "tags": [],
            "needs_background": "light"
          }
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "hAWVK8PAfW2F",
        "outputId": "d51067c8-f752-461d-e36e-39530d37c393"
      },
      "source": [
        "average_qual[66]"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "4.526"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 106
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "3LzCYzUIlkKH"
      },
      "source": [
        "# Week3 \n",
        "\n",
        "## Hamming Distance and edit distance\n",
        "\n",
        "editDistance()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "06M9PQI-l8K5"
      },
      "source": [
        "\n",
        "## Read alignment in the field\n",
        "\n",
        "Index acts as a filtering. \n",
        "\n",
        "\n",
        "Small set of needdles."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "2xxCjac2mX-S"
      },
      "source": [
        "Index and dynamic programming cover each other's flaw.\n",
        "\n",
        "Index doesn't deal with mismatches\n",
        "\n",
        "Dynaminal program cann't handle big matrix.\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ByT5JBcVnPov"
      },
      "source": [
        "##  Assembly: working from scratch\n",
        "\n",
        "\n",
        "\n",
        "### Denovo assembly\n",
        "\n",
        "![image](https://user-images.githubusercontent.com/20909751/100520200-928de800-316a-11eb-9916-fe6966b7e7c6.png)\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "SX5n-J47oKOQ"
      },
      "source": [
        "Coverage\n",
        "\n",
        "![image](https://user-images.githubusercontent.com/20909751/100520265-ec8ead80-316a-11eb-9b13-db76e4b0ce0e.png)\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "bbNwGn4spAfA"
      },
      "source": [
        "Average coverage\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "S5LF0oV_qKeR"
      },
      "source": [
        "* First law of assembly\n",
        "\n",
        "If a suffix of read A is smilar to prefix of read B, then A and B might overlap in the genome\n",
        "\n",
        "* Second law of assembly\n",
        "\n",
        "More coverage leads to more and long overlaps\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "85uG8QBOqKst"
      },
      "source": [
        "### Practical: Overlaps between pairs of reads\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "WY3KpTajs3yO"
      },
      "source": [
        "def overlap(a, b, min_length=3):\n",
        "    \"\"\"\n",
        "\n",
        "    \"\"\"\n",
        "    start = 0\n",
        "    while True:\n",
        "        start = a.find(b[:min_length], start)\n",
        "        if start == -1:\n",
        "            return 0\n",
        "        if b.startswith(a[start:]):\n",
        "            return len(a) - start\n",
        "        start += 1\n",
        "\n"
      ],
      "execution_count": 1,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "GA7-1sq2tzSQ",
        "outputId": "7de7b1f9-51a8-4c2d-f765-c02a69e34521"
      },
      "source": [
        "## Overlap with 3 letters\n",
        "overlap(\"TTACGT\", \"CGTACCGT\")"
      ],
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "3"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 2
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "fiH1Z2Eat6-2",
        "outputId": "4bcc5d59-e2d3-4c5c-e50c-4416ee7bf8ee"
      },
      "source": [
        "## Overlap with two letters\n",
        "\n",
        "overlap(\"TTACGT\", \"GTACCGT\")"
      ],
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "0"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 3
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "mwLOEafhuK2L"
      },
      "source": [
        "### Practical: Finding and representing all overlaps\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "s1bsEKl8u156",
        "outputId": "e9f7ac57-fcd8-478b-9fa1-a766acb608ac"
      },
      "source": [
        "from itertools import permutations\n",
        "list(permutations([1,2,3], 3))"
      ],
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "[(1, 2, 3), (1, 3, 2), (2, 1, 3), (2, 3, 1), (3, 1, 2), (3, 2, 1)]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 4
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "QNLJrDWsvBC7"
      },
      "source": [
        "def naive_overlap_map(reads, k):\n",
        "    olaps = {}\n",
        "    for a, b in permutations(reads, 2):\n",
        "      olen = overlap(a, b, min_length=k)\n",
        "      if olen > 0:\n",
        "          olaps[(a, b)] = olen\n",
        "    return olaps\n"
      ],
      "execution_count": 7,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "apkXqTjNvjmO",
        "outputId": "96bf2894-0f89-47e2-899d-e35f3aab0e24"
      },
      "source": [
        "reads = ['ACGGACGATC', \"GATCAAGT\", \"TTCACGGA\"]\n",
        "print(naive_overlap_map(reads, 3))"
      ],
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "{('ACGGACGATC', 'GATCAAGT'): 4, ('TTCACGGA', 'ACGGACGATC'): 5}\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "AKyTvs91wDYm"
      },
      "source": [
        "# Week 4\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "RQX4HMJIwzyj"
      },
      "source": [
        "## The shortest common superstring problem\n",
        "\n",
        "Given as a set of strings __*S*__, find SCS (S): shortest string contaming the strings in __*S*__ "
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "jzfYPShuQCi2"
      },
      "source": [
        ""
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "PZYmth1zQDEG"
      },
      "source": [
        "Idea: pick order for strings in __*S*__ and construct superstring.\n",
        "\n",
        "Try all posible orderings and pcik shorted superstring\n",
        "\n",
        "If $S$ contains $n$ strings, $n!$ (n factoirial) orderings possible. \n",
        "\n",
        "So n factorial is a quantity that grows very quickly as $n$ grows. As the number of input strings grows, the amount of work that we have to do, the amount of permutations we have to try grows very, very rapidly. That's why we call this problem an intractable problem. "
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "ppYtapV4wGx5"
      },
      "source": [
        "\n",
        "import itertools\n",
        "\n",
        "def scs(ss):\n",
        "    \"\"\"\n",
        "    brute force method????\n",
        "    Returns shortest common superstring of given strings,\n",
        "    assuming no string is a strict substring of another \n",
        "    \"\"\"\n",
        "    shortest_sup = None\n",
        "    ## return every possible orders\n",
        "    for ssperm in itertools.permutations(ss):\n",
        "        sup = ssperm[0]\n",
        "        for i in range(len(ss)-1):\n",
        "            olen = overlap(ssperm[i], ssperm[i+1], min_length=1)\n",
        "            sup += ssperm[i+1][olen:]\n",
        "        if shortest_sup is None or len(sup) < len(shortest_sup):\n",
        "            shortest_sup = sup\n",
        "    return shortest_sup"
      ],
      "execution_count": 9,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 35
        },
        "id": "6No48hCQPfUG",
        "outputId": "5be848d7-0369-46ba-c9af-7cc0981559bc"
      },
      "source": [
        "scs(['ACGGATGAGC', 'GAGCGGA', 'GAGCGAG'])"
      ],
      "execution_count": 11,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            },
            "text/plain": [
              "'ACGGATGAGCGAGCGGA'"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 11
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "CuWq8EWcWDD0"
      },
      "source": [
        "###  Greedy shortest common superstring\n",
        "\n",
        "Picking the longest overlap seems to make sense because the longer the overlaps between the strings, the shorter the final string will be. So that's what we mean by greedy algorithm in this case. \n",
        "\n",
        "The greedy alogrith doesn't always give the correct results. \n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "K6QyV33RYYAV"
      },
      "source": [
        "import itertools\n",
        "def pick_maximal_overlap(reads, k):\n",
        "    reada, readb = None, None\n",
        "    best_olen = 0\n",
        "    for a, b in itertools.permutations(reads, 2):\n",
        "        olen = overlap(a, b, min_length=k)\n",
        "        # There might be multiple pairs that are tied for the maximal overlap length\n",
        "        if olen > best_olen:\n",
        "            reada, readb = a, b\n",
        "            best_olen = olen\n",
        "    return reada, readb, best_olen\n",
        "\n",
        "def greedy_scs(reads, k):\n",
        "    readA, readB, olen = pick_maximal_overlap(reads, k)\n",
        "    while olen > 0:\n",
        "        reads.remove(readA)\n",
        "        reads.remove(readB)\n",
        "        reads.append(readA+readB[olen:])\n",
        "        readA, readB, olen = pick_maximal_overlap(reads, k)\n",
        "    # we merge and merge and merge we might get to the point where there are no\n",
        "    # more edges, no more overlaps left in the overlap graph, but there are still \n",
        "    #multiple nodes left, in which case we can just concatenate the node labels \n",
        "    # together to get our super string. \n",
        "    return ''.join(reads)"
      ],
      "execution_count": 20,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 35
        },
        "id": "tSFxVYafa28v",
        "outputId": "22489ac2-4b03-40f1-a5dc-fdc68e7ce9d5"
      },
      "source": [
        "greedy_scs(['ABC', \"BCA\", \"CAB\"], 2)"
      ],
      "execution_count": 21,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            },
            "text/plain": [
              "'CABCA'"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 21
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 35
        },
        "id": "RKg8GtiNbOPv",
        "outputId": "18143bff-f484-4a3c-acc0-2a5864d3f73e"
      },
      "source": [
        "greedy_scs(['ABCD', \"CDBC\", \"BCDA\"], 1)"
      ],
      "execution_count": 22,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            },
            "text/plain": [
              "'CDBCABCDA'"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 22
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 35
        },
        "id": "BG9__UEvbYoA",
        "outputId": "a441988d-921a-41e7-e308-e875ec42d7f2"
      },
      "source": [
        "# Let's run it with the brute force method and see what the best one for that is. \n",
        "# So you can see on this that the shortest common superstring is actually shorter \n",
        "# than the Greedy superstring, and the reason for this is to get the shortest common\n",
        "# superstring, you can just combine these reads in order, and each one overlaps by \n",
        "# two with the one after it. \n",
        "# But the first string and the third string actually overlap by three. So the Greedy \n",
        "# algorithm will initially merge those two together into a larger string, and \n",
        "# then it has no option but to just append the middle string to the end since it doesn't \n",
        "# have any overlap. \n",
        "# So it actually does a bit worse by doing the Greedy method here. #\n",
        "# But that trade-off in terms of accuracy, in most cases, will be worth the time \n",
        "# you save by doing the Greedy method right? \n",
        "\n",
        "scs(['ABCD', \"CDBC\", \"BCDA\"])"
      ],
      "execution_count": 25,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            },
            "text/plain": [
              "'ABCDBCDA'"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 25
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "AgQ9LhUqb-4y"
      },
      "source": [
        "### Third law of assembly: repeats are bad\n",
        "\n",
        "Repetitive sequences can make it difficult or even impossible to correctly assemble the original genome. Repetitive sequences cause ambiguity, and just like if you're trying to put together a puzzle and the puzzle is like half featureless sky pieces, just like in that case, it's not always going to be easy or possible to resolve the ambiguity. So the third law of assembly states, repeats make assembly difficult. \n",
        "\n",
        "called the de Bruijn graph and the Eulerian walk method.\n",
        "Play video starting at :5:31 and follow transcript5:31\n",
        "That's an alternative that actually avoids the over collapsing problem. So it avoids this particular problem. But it still can't avoid the overall problem that repeats make assembly difficult. And in another later lecture, we'll see how real assembly software deals with this issue. And then in another lecture, we'll look a bit closer at whether new sequencing technologies might be able to help us sidestep this problem.\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "dJJFbPYndecM"
      },
      "source": [
        "### Building a De Bruijn graph\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "HHEgNQoRdbu3"
      },
      "source": [
        "def de_bruijn_ize(st, k):\n",
        "    edges = []\n",
        "    nodes = set()\n",
        "    for i in range(len(st) - k +1 ):\n",
        "      edges.append((st[i:i+k-1], st[i+1:i+k]))\n",
        "      nodes.add(st[i:i+k-1])\n",
        "      nodes.add(st[i+1:i+k])\n",
        "    return nodes, edges  "
      ],
      "execution_count": 54,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "2aDulQSZeQ3E",
        "outputId": "ecd9c245-836c-4cb4-9c5d-d81f5ee52cef"
      },
      "source": [
        "nodes, edges = de_bruijn_ize(\"ACGCGTCG\", 3)\n",
        "print(nodes)\n",
        "print(edges)"
      ],
      "execution_count": 55,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "{'GT', 'CG', 'AC', 'TC', 'GC'}\n",
            "[('AC', 'CG'), ('CG', 'GC'), ('GC', 'CG'), ('CG', 'GT'), ('GT', 'TC'), ('TC', 'CG')]\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "m3SsfQ_dez6X"
      },
      "source": [
        "def visualize_de_bruijn(st, k):\n",
        "    \"\"\" Visualize a directed multigraph using graphviz \"\"\"\n",
        "    nodes, edges = de_bruijn_ize(st, k)\n",
        "    dot_str = 'digraph \"DeBruijn graph\" {\\n'\n",
        "    for node in nodes:\n",
        "        dot_str += '  %s [label=\"%s\"] ;\\n' % (node, node)\n",
        "    for src, dst in edges:\n",
        "        dot_str += '  %s -> %s ;\\n' % (src, dst)\n",
        "    return dot_str + '}\\n'"
      ],
      "execution_count": 56,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Qs-UgYzbfVt1",
        "outputId": "bab9d80d-d2ec-4e7a-ff08-41c4c0e58704"
      },
      "source": [
        "!pip install gvmagic"
      ],
      "execution_count": 37,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Collecting gvmagic\n",
            "  Downloading https://files.pythonhosted.org/packages/f7/f8/725b48babca08b58e863bd6f2335641578874c795c5936f8aba1802b74a9/gvmagic-0.5.tar.gz\n",
            "Building wheels for collected packages: gvmagic\n",
            "  Building wheel for gvmagic (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for gvmagic: filename=gvmagic-0.5-cp36-none-any.whl size=10376 sha256=7fb1c81be490dd75f352c746c2262a40f7dfaa2e900fb6404279d41044c1d391\n",
            "  Stored in directory: /root/.cache/pip/wheels/f6/67/1f/64d62350d4cffe0a94e39af37828a6a5bfba1b9b9d6d95dbf8\n",
            "Successfully built gvmagic\n",
            "Installing collected packages: gvmagic\n",
            "Successfully installed gvmagic-0.5\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "ACCCFZNTfWow"
      },
      "source": [
        "%load_ext gvmagic"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 368
        },
        "id": "jaBVCYEVe0At",
        "outputId": "2c5a1d04-f48b-40fe-80b6-d1f351a75e06"
      },
      "source": [
        "%dotstr visualize_de_bruijn(\"ACGCGTCG\", 3)\n"
      ],
      "execution_count": 57,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.SVG object>"
            ],
            "image/svg+xml": "<svg height=\"260pt\" viewBox=\"0.00 0.00 172.00 260.00\" width=\"172pt\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n<g class=\"graph\" id=\"graph0\" transform=\"scale(1 1) rotate(0) translate(4 256)\">\n<title>DeBruijn graph</title>\n<polygon fill=\"#ffffff\" points=\"-4,4 -4,-256 168,-256 168,4 -4,4\" stroke=\"transparent\"/>\n<!-- GT -->\n<g class=\"node\" id=\"node1\">\n<title>GT</title>\n<ellipse cx=\"54\" cy=\"-234\" fill=\"none\" rx=\"27\" ry=\"18\" stroke=\"#000000\"/>\n<text fill=\"#000000\" font-family=\"Times,serif\" font-size=\"14.00\" text-anchor=\"middle\" x=\"54\" y=\"-230.3\">GT</text>\n</g>\n<!-- TC -->\n<g class=\"node\" id=\"node4\">\n<title>TC</title>\n<ellipse cx=\"27\" cy=\"-162\" fill=\"none\" rx=\"27\" ry=\"18\" stroke=\"#000000\"/>\n<text fill=\"#000000\" font-family=\"Times,serif\" font-size=\"14.00\" text-anchor=\"middle\" x=\"27\" y=\"-158.3\">TC</text>\n</g>\n<!-- GT&#45;&gt;TC -->\n<g class=\"edge\" id=\"edge5\">\n<title>GT-&gt;TC</title>\n<path d=\"M47.3258,-216.2022C44.2524,-208.0064 40.5384,-198.1024 37.1305,-189.0145\" fill=\"none\" stroke=\"#000000\"/>\n<polygon fill=\"#000000\" points=\"40.3858,-187.7274 33.5974,-179.593 33.8315,-190.1853 40.3858,-187.7274\" stroke=\"#000000\"/>\n</g>\n<!-- CG -->\n<g class=\"node\" id=\"node2\">\n<title>CG</title>\n<ellipse cx=\"82\" cy=\"-90\" fill=\"none\" rx=\"27\" ry=\"18\" stroke=\"#000000\"/>\n<text fill=\"#000000\" font-family=\"Times,serif\" font-size=\"14.00\" text-anchor=\"middle\" x=\"82\" y=\"-86.3\">CG</text>\n</g>\n<!-- CG&#45;&gt;GT -->\n<g class=\"edge\" id=\"edge4\">\n<title>CG-&gt;GT</title>\n<path d=\"M78.4664,-108.1727C73.6642,-132.8697 65.0329,-177.2595 59.4117,-206.1684\" fill=\"none\" stroke=\"#000000\"/>\n<polygon fill=\"#000000\" points=\"55.9627,-205.5692 57.4896,-216.0535 62.834,-206.9054 55.9627,-205.5692\" stroke=\"#000000\"/>\n</g>\n<!-- GC -->\n<g class=\"node\" id=\"node5\">\n<title>GC</title>\n<ellipse cx=\"82\" cy=\"-18\" fill=\"none\" rx=\"27\" ry=\"18\" stroke=\"#000000\"/>\n<text fill=\"#000000\" font-family=\"Times,serif\" font-size=\"14.00\" text-anchor=\"middle\" x=\"82\" y=\"-14.3\">GC</text>\n</g>\n<!-- CG&#45;&gt;GC -->\n<g class=\"edge\" id=\"edge2\">\n<title>CG-&gt;GC</title>\n<path d=\"M76.0843,-72.2022C75.2798,-64.1807 75.0571,-54.5231 75.4162,-45.5957\" fill=\"none\" stroke=\"#000000\"/>\n<polygon fill=\"#000000\" points=\"78.9097,-45.8099 76.105,-35.593 71.9262,-45.3289 78.9097,-45.8099\" stroke=\"#000000\"/>\n</g>\n<!-- AC -->\n<g class=\"node\" id=\"node3\">\n<title>AC</title>\n<ellipse cx=\"137\" cy=\"-162\" fill=\"none\" rx=\"27\" ry=\"18\" stroke=\"#000000\"/>\n<text fill=\"#000000\" font-family=\"Times,serif\" font-size=\"14.00\" text-anchor=\"middle\" x=\"137\" y=\"-158.3\">AC</text>\n</g>\n<!-- AC&#45;&gt;CG -->\n<g class=\"edge\" id=\"edge1\">\n<title>AC-&gt;CG</title>\n<path d=\"M124.5196,-145.6621C117.4876,-136.4564 108.5653,-124.7764 100.7117,-114.4953\" fill=\"none\" stroke=\"#000000\"/>\n<polygon fill=\"#000000\" points=\"103.2828,-112.0953 94.431,-106.2733 97.72,-116.3446 103.2828,-112.0953\" stroke=\"#000000\"/>\n</g>\n<!-- TC&#45;&gt;CG -->\n<g class=\"edge\" id=\"edge6\">\n<title>TC-&gt;CG</title>\n<path d=\"M39.4804,-145.6621C46.5124,-136.4564 55.4347,-124.7764 63.2883,-114.4953\" fill=\"none\" stroke=\"#000000\"/>\n<polygon fill=\"#000000\" points=\"66.28,-116.3446 69.569,-106.2733 60.7172,-112.0953 66.28,-116.3446\" stroke=\"#000000\"/>\n</g>\n<!-- GC&#45;&gt;CG -->\n<g class=\"edge\" id=\"edge3\">\n<title>GC-&gt;CG</title>\n<path d=\"M87.895,-35.593C88.7094,-43.5865 88.942,-53.2363 88.5928,-62.1759\" fill=\"none\" stroke=\"#000000\"/>\n<polygon fill=\"#000000\" points=\"85.0976,-61.9891 87.9157,-72.2022 92.0817,-62.4608 85.0976,-61.9891\" stroke=\"#000000\"/>\n</g>\n</g>\n</svg>"
          },
          "metadata": {
            "tags": []
          }
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 368
        },
        "id": "jnDVDdrvez-D",
        "outputId": "7b65fd00-8ffc-4d7c-86eb-23846297123b"
      },
      "source": [
        "%dotstr visualize_de_bruijn(\"AAABBBBA\", 3)\n"
      ],
      "execution_count": 58,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.SVG object>"
            ],
            "image/svg+xml": "<svg height=\"260pt\" viewBox=\"0.00 0.00 98.00 260.00\" width=\"98pt\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n<g class=\"graph\" id=\"graph0\" transform=\"scale(1 1) rotate(0) translate(4 256)\">\n<title>DeBruijn graph</title>\n<polygon fill=\"#ffffff\" points=\"-4,4 -4,-256 94,-256 94,4 -4,4\" stroke=\"transparent\"/>\n<!-- BB -->\n<g class=\"node\" id=\"node1\">\n<title>BB</title>\n<ellipse cx=\"27\" cy=\"-90\" fill=\"none\" rx=\"27\" ry=\"18\" stroke=\"#000000\"/>\n<text fill=\"#000000\" font-family=\"Times,serif\" font-size=\"14.00\" text-anchor=\"middle\" x=\"27\" y=\"-86.3\">BB</text>\n</g>\n<!-- BB&#45;&gt;BB -->\n<g class=\"edge\" id=\"edge4\">\n<title>BB-&gt;BB</title>\n<path d=\"M52.4435,-96.6907C63.0278,-97.1521 72,-94.9219 72,-90 72,-86.77 68.136,-84.6992 62.4881,-83.7876\" fill=\"none\" stroke=\"#000000\"/>\n<polygon fill=\"#000000\" points=\"62.5987,-80.289 52.4435,-83.3093 62.2656,-87.2811 62.5987,-80.289\" stroke=\"#000000\"/>\n</g>\n<!-- BB&#45;&gt;BB -->\n<g class=\"edge\" id=\"edge5\">\n<title>BB-&gt;BB</title>\n<path d=\"M48.6268,-101.0075C68.5379,-106.5363 90,-102.8672 90,-90 90,-79.2438 75.0025,-74.9152 58.4504,-77.014\" fill=\"none\" stroke=\"#000000\"/>\n<polygon fill=\"#000000\" points=\"57.7389,-73.587 48.6268,-78.9925 59.1211,-80.4492 57.7389,-73.587\" stroke=\"#000000\"/>\n</g>\n<!-- BA -->\n<g class=\"node\" id=\"node4\">\n<title>BA</title>\n<ellipse cx=\"27\" cy=\"-18\" fill=\"none\" rx=\"27\" ry=\"18\" stroke=\"#000000\"/>\n<text fill=\"#000000\" font-family=\"Times,serif\" font-size=\"14.00\" text-anchor=\"middle\" x=\"27\" y=\"-14.3\">BA</text>\n</g>\n<!-- BB&#45;&gt;BA -->\n<g class=\"edge\" id=\"edge6\">\n<title>BB-&gt;BA</title>\n<path d=\"M27,-71.8314C27,-64.131 27,-54.9743 27,-46.4166\" fill=\"none\" stroke=\"#000000\"/>\n<polygon fill=\"#000000\" points=\"30.5001,-46.4132 27,-36.4133 23.5001,-46.4133 30.5001,-46.4132\" stroke=\"#000000\"/>\n</g>\n<!-- AA -->\n<g class=\"node\" id=\"node2\">\n<title>AA</title>\n<ellipse cx=\"27\" cy=\"-234\" fill=\"none\" rx=\"27\" ry=\"18\" stroke=\"#000000\"/>\n<text fill=\"#000000\" font-family=\"Times,serif\" font-size=\"14.00\" text-anchor=\"middle\" x=\"27\" y=\"-230.3\">AA</text>\n</g>\n<!-- AA&#45;&gt;AA -->\n<g class=\"edge\" id=\"edge1\">\n<title>AA-&gt;AA</title>\n<path d=\"M46.895,-246.4321C59.688,-249.6753 72,-245.5313 72,-234 72,-225.6218 65.5006,-221.1433 57.0395,-220.5644\" fill=\"none\" stroke=\"#000000\"/>\n<polygon fill=\"#000000\" points=\"56.5019,-217.1004 46.895,-221.5679 57.191,-224.0665 56.5019,-217.1004\" stroke=\"#000000\"/>\n</g>\n<!-- AB -->\n<g class=\"node\" id=\"node3\">\n<title>AB</title>\n<ellipse cx=\"27\" cy=\"-162\" fill=\"none\" rx=\"27\" ry=\"18\" stroke=\"#000000\"/>\n<text fill=\"#000000\" font-family=\"Times,serif\" font-size=\"14.00\" text-anchor=\"middle\" x=\"27\" y=\"-158.3\">AB</text>\n</g>\n<!-- AA&#45;&gt;AB -->\n<g class=\"edge\" id=\"edge2\">\n<title>AA-&gt;AB</title>\n<path d=\"M27,-215.8314C27,-208.131 27,-198.9743 27,-190.4166\" fill=\"none\" stroke=\"#000000\"/>\n<polygon fill=\"#000000\" points=\"30.5001,-190.4132 27,-180.4133 23.5001,-190.4133 30.5001,-190.4132\" stroke=\"#000000\"/>\n</g>\n<!-- AB&#45;&gt;BB -->\n<g class=\"edge\" id=\"edge3\">\n<title>AB-&gt;BB</title>\n<path d=\"M27,-143.8314C27,-136.131 27,-126.9743 27,-118.4166\" fill=\"none\" stroke=\"#000000\"/>\n<polygon fill=\"#000000\" points=\"30.5001,-118.4132 27,-108.4133 23.5001,-118.4133 30.5001,-118.4132\" stroke=\"#000000\"/>\n</g>\n</g>\n</svg>"
          },
          "metadata": {
            "tags": []
          }
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "O14WAvsVQ9w3"
      },
      "source": [
        "# Reference\n",
        "\n",
        "\n",
        "https://github.com/sidsriv/Algorithms-for-DNA-sequencing\n",
        "\n",
        "https://github.com/pejvan/Algorithms-for-DNA-Sequencing\n",
        "\n",
        "https://github.com/claytonjwong/Algorithms-DNA-Sequencing\n",
        "\n",
        "https://github.com/kritikadusad/AlgorithmsForDNASequencing\n",
        "\n",
        "https://www.sciencedirect.com/science/article/abs/pii/S0377221716304830\n",
        "\n",
        "Dimitri J. 2020. Exact and approximation algorithms for DNA sequence reconstruction.\n",
        "\n",
        "Najam M, Rasool RU, Ahmad HF, Ashraf U, Malik AW. 2019. Pattern Matching for DNA Sequencing Data Using Multiple Bloom Filters. Biomed Res Int 2019. \n",
        "\n",
        "https://github.com/mrhua2019/Pattern-Matching-for-DNA-Sequencing-Data-using-Multiple-Bloom-Filters#readme\n",
        "\n",
        "Langmead B. 2557. ALGORITHMS AND HIGH PERFORMANCE COMPUTING APPROACHES FOR SEQUENCING-BASED COMPARATIVE GENOMICS. 7: 1–16. https://drum.lib.umd.edu/handle/1903/12569.\n"
      ]
    }
  ]
}